File: app/src/main/java/com/saladevs/changelogclone/ui/BasePresenter.java
package com.saladevs.changelogclone.ui;

import android.support.annotation.Nullable;

public abstract class BasePresenter<T extends MvpView> {
    private T mMvpView;

    public void attachView(T mvpView) {
        mMvpView = mvpView;
    }

    public void detachView() {
        mMvpView = null;
    }

    private boolean isViewAttached() {
        return mMvpView != null;
    }

    @Nullable
    protected T getMvpView() {
        return mMvpView;
    }

    protected void checkViewAttached() {
        if (!isViewAttached()) throw new MvpViewNotAttachedException();
    }

    private static class MvpViewNotAttachedException extends RuntimeException {
        public MvpViewNotAttachedException() {
            super("Please call Presenter.attachView(MvpView) before" +
                    " requesting data to the Presenter");
        }
    }
}

File: app/src/main/java/com/saladevs/changelogclone/ui/BasePresenter.kt
package com.saladevs.changelogclone.ui

import io.realm.Realm
import rx.subscriptions.CompositeSubscription

abstract class BasePresenter<T : MvpView> {

    protected val realm = Realm.getDefaultInstance()
    protected val subscriptions = CompositeSubscription()

    protected var mvpView: T? = null
        private set

    open fun attachView(mvpView: T) {
        this.mvpView = mvpView
    }

    open fun detachView() {
        subscriptions.clear()
        realm.close()
        mvpView = null
    }

    private val isViewAttached: Boolean
        get() = mvpView != null

    protected fun checkViewAttached() {
        if (!isViewAttached) throw MvpViewNotAttachedException()
    }

    private class MvpViewNotAttachedException : RuntimeException("Please call Presenter.attachView(MvpView) before" + " requesting data to the Presenter")
}
